internal/selectors/selectors.go

```diff
@@ -15,23 +15,39 @@ import (
 	api "sigs.k8s.io/hierarchical-namespaces/api/v1alpha2"
 )
 
-func ShouldPropagate(inst *unstructured.Unstructured, nsLabels labels.Set) (bool, error) {
-	if sel, err := GetSelector(inst); err != nil {
+// ShouldPropagate returns true if the given object should be propagated
+// based on the SynchronizationMode and on the selectors of the object.
+// Propagation will be done only with 'Propagate' or 'AllowPropagate' modes.
+// When selectors are set, 'AllowPropagate' follows the same logic as 'Propagate' mode,
+// when they are not set, then if 'Propagate' mode is used then propagate by default;
+// if 'AllowPropagate' mode is used then do not propagate by default.
+func ShouldPropagate(inst *unstructured.Unstructured, nsLabels labels.Set, mode api.SynchronizationMode) (bool, error) {
+	sel, err := GetSelector(inst)
+	if err != nil {
 		return false, err
 	} else if sel != nil && !sel.Matches(nsLabels) {
```

Comment:
No need for `else` here since `sel` is now defined outside of the if statement. Alternatively, maybe something like this would be easier to read:

```
propIfNotExcluded := (mode == api.Propagate)

if sel, err := GetSelector(inst); err != nil {
  return false, err
} else if sel != nil {
  propIfNotExcluded = true
  if !sel.Matches(nsLabels) {
    return false, nil
  }
}
```

Then at the end of the function, it's just `return propIfNotExcluded`. Up to you if you think that's clearer though, I know those nested `if` statements are a pain.

Reply:
Think it's clearer yes, need to check if `!sel.Empty()` as well.